kvserver: auto-tune closed timestamp updates

wenyihu6 opened the pull request:
**Note that this PR is based on https://github.com/cockroachdb/cockroach/pull/142281.**

---

**kvserver: auto-tune closed timestamp updates**

Previously, for global reads, we calculated the target closed timestamp using
the `LEAD_FOR_GLOBAL_READS` policy. This policy indicates that the range’s closed
timestamp is configured to lead the present time so that all follower replicas
of the range can serve present time reads. Note that this is an estimate/goal
but not the lead time itself.

This is how the calculation is currently done:

```
raft_propagation_time = max_network_rtt*1.5 (closed timestamp propagation by raft from a leader)
+ raft_overhead(proposal evaluation, log sync, and state machine apply)

side_propagation_time = max_network_rtt*0.5 (closed timestamp propagation by side transport)
+ side_transport_close_interval

// Maximum time required to propagate a closed timestamp update to followers from the sender.
closed_ts_propagation = max(raft_propagation_time, side_propagation_time)

closed_ts_at_sender = now(sender’s current clock) + maxClockOffset (max clock uncertainty)
+ closed_ts_propagation
```

max_network_rtt is a hardcoded 150ms. By default, this results in a closed timestamp leading the sender's current clock by 800ms. Users can override this by tuning `kv.closed_timestamp.lead_for_global_reads_override`.

**Problems with the previous approach**

1. Tuning trade-offs: A higher lead timestamp forces leaseholders to wait longer
for the current time to catch up, incurring a high write latency. A lower lead
timestamp may result in followers not fully replicating the changes before
returning control to the client, forcing follower replicas to redirect query to
the leaseholder and increase read latency.

2. Cluster wide setting: This setting applies to the entire cluster, not per
range. For geographically distant replicas, a short lead time may not be enough
for full application. However, increasing the lead time cluster wide can harm
write performance for replicas that are close together and don’t require as much
time to replicate.

**Improvements in this commit**

This commit improves closed timestamp computation for global reads by using actual network
latency measurements instead of hardcoded values.

Alternatives considered:

1. Ideally, we would account for the full application time on follower replicas.
However, the leader has no direct visibility into when follower replicas apply
updates to their state machines.

2. Ideally, every node-to-node RPC would have its own network latency average,
resulting in a unique closed timestamp target for each pair. However, this would
significantly complicate the side transport design. Currently, senders and
receivers share state using a single closed timestamp per policy, with any new
or removed ranges recorded in the group. To introduce unique timestamps per
range based on varying network latencies, we would need to implement a
node-to-node latency mapping on top of the existing group ranges state view.

Beyond this added complexity, we must also account for the maximum of the two
propagation times. This requires side transport closed timestamp computations to
be aware of raft propagation delays, and vice versa. However, we currently lack
the necessary infrastructure to share this information between replicas and side
transport. Given that this is only a target, introducing such infrastructure
does not seem justified due to its complexity. Notes: this is just a target
computation. Some follower reads may still occur, particularly if we fail to
update closed timestamps on the side transport or if certain follower replicas
lag behind and fail to catch up quickly.

Resolves: https://github.com/cockroachdb/cockroach/issues/59680
Release note: `kv.closed_timestamp.lead_for_global_reads_auto_tune.enabled`
can now be used to auto-tune the lead time that ranges global tables use to publish
close timestamps. The `kv.closed_timestamp.lead_for_global_reads_override`
cluster setting takes precedence over this one. If auto-tuning is disabled or
no data is observed, the system falls back to a hardcoded computed lead time.

---

**kvserver: cache network RTT for closed timestamp target computation**

Previously, during every FlushLockedWithRaftGroup call, we would compute the
maximum network RTT by iterating over all replicas and querying their latencies
from the centralized RPC context. This was expensive as it required accessing
the shared RemoteClocks.Latency callback on this hot path.

This commit introduces a time-based cache for the maximum RTT value. The cache
is refreshed every 5 minutes for stable values, or every 10 seconds if the
previous value was outside acceptable bounds (0-400ms). While this may lead to
slightly less accurate closed timestamp target computations, the performance
benefit outweighs the minor precision loss.

Part of: https://github.com/cockroachdb/cockroach/issues/59680
Release note: none

---

**kvserver: use randomized refresh interval to refresh maxReplicaRTT**

Previously, all ranges on a node shared the same refresh interval, leading to
simultaneous accesses to the RPC context clock and causing contentions. This
commit introduces a randomized jitter to the refresh interval, ensuring
different ranges on different stores of the same node have varying intervals,
reducing contention and improving performance.

Part of: #59680
Release note: none

wenyihu6 committed:
kvserver/closedts: use max() instead of conditional checks

Previously, conditional check is used to find the maximum
between sideTransportPropTime and raftTransportPropTime.

This patch changes it to use the max builtin for better
readability.

Release note: none
Epic: none

wenyihu6 committed:
kvserver/closedts: move leadTargetOverride above for better readability

Previously, if a cluster defines `kv.closed_timestamp.lead_for_global_reads_override`,
the lead time calclulated at the sender will just be overriden in the end.

This patch moves the override part earlier in the logic and
breaks on an early return for better readability.

Epic: none
Release note: none

wenyihu6 committed:
kvserver: add kv.closed_timestamp.lead_for_global_reads_auto_tune.enabled

This patch introduces a new cluster setting,
`kv.closed_timestamp.lead_for_global_reads_auto_tune.enabled`,
which enables auto tuning of the lead time that global_read ranges use to
publish closed timestamps. The tuning is based on observed raft and
side-transport closed timestamp propagation latencies.
The `kv.closed_timestamp.lead_for_global_reads_override` cluster setting
takes precedence over this one. If auto-tuning is disabled or no data is
observed, the system falls back to a hardcoded computed lead time.

Note that setting is currently unused, and no release notes have been added yet.
Future commits will populate and integrate it.

Part of: #59680
Release note: none

wenyihu6 committed:
kvserver/closedts: add avgMaxNetWorkLatency to sidetransport.Sender

This patch adds a new field `avgMaxNetWorkLatency` to Sender, tracking the
average maximum network latency between the sender and any of its follower nodes
An exponentially weighted moving average is used with an effective window size
of 30, where recent data points have a higher influence, and older data
gradually decays.

Note that `avgMaxNetWorkLatency` is left unused. Future commits will use it
to dyanamically compute closed timestamp target for global tables.

Part of: #59680
Release note: none

wenyihu6 committed:
kvserver/closedts: extract part of TargetForPolicy

This patch refactors `closedts.TargetForPolicy` by extracting part of its code
into a helper function. It does not change any existing behavior but makes
future commits cleaner.

Part of: #59680
Release note: none

wenyihu6 committed:
kvserver: auto-tune closed timestamp updates

Previously, for global reads, we calculated the target closed timestamp using
the `LEAD_FOR_GLOBAL_READS` policy. This policy indicates that the range’s closed
timestamp is configured to lead the present time so that all follower replicas
of the range can serve present time reads. Note that this is an estimate/goal
but not the lead time itself.

This is how the calculation is currently done:

```
raft_propagation_time = max_network_rtt*1.5 (closed timestamp propagation by raft from a leader)
+ raft_overhead(proposal evaluation, log sync, and state machine apply)

side_propagation_time = max_network_rtt*0.5 (closed timestamp propagation by side transport)
+ side_transport_close_interval

// Maximum time required to propagate a closed timestamp update to followers from the sender.
closed_ts_propagation = max(raft_propagation_time, side_propagation_time)

closed_ts_at_sender = now(sender’s current clock) + maxClockOffset (max clock uncertainty)
+ closed_ts_propagation
```

max_network_rtt is a hardcoded 150ms. By default, this results in a closed timestamp leading
the sender's current clock by 800ms. Users can override this by tuning
`kv.closed_timestamp.lead_for_global_reads_override`.

**Problems with the previous approach**

1. Tuning trade-offs: A higher lead timestamp forces leaseholders to wait longer
for the current time to catch up, incurring a high write latency. A lower lead
timestamp may result in followers not fully replicating the changes before
returning control to the client, forcing follower replicas to redirect query to
the leaseholder and increase read latency.

2. Cluster wide setting: This setting applies to the entire cluster, not per
range. For geographically distant replicas, a short lead time may not be enough
for full application. However, increasing the lead time cluster wide can harm
write performance for replicas that are close together and don’t require as much
time to replicate.

**Improvements in this commit**

This commit improves closed timestamp computation for global reads by using actual network
latency measurements instead of hardcoded values.

Alternatives considered:

1. Ideally, we would account for the full application time on follower replicas.
However, the leader has no direct visibility into when follower replicas apply
updates to their state machines.

2. Ideally, every node-to-node RPC would have its own network latency average,
resulting in a unique closed timestamp target for each pair. However, this would
significantly complicate the side transport design. Currently, senders and
receivers share state using a single closed timestamp per policy, with any new
or removed ranges recorded in the group. To introduce unique timestamps per
range based on varying network latencies, we would need to implement a
node-to-node latency mapping on top of the existing group ranges state view.

Beyond this added complexity, we must also account for the maximum of the two
propagation times. This requires side transport closed timestamp computations to
be aware of raft propagation delays, and vice versa. However, we currently lack
the necessary infrastructure to share this information between replicas and side
transport. Given that this is only a target, introducing such infrastructure
does not seem justified due to its complexity. Notes: this is just a target
computation. Some follower reads may still occur, particularly if we fail to
update closed timestamps on the side transport or if certain follower replicas
lag behind and fail to catch up quickly.

Resolves: https://github.com/cockroachdb/cockroach/issues/59680
Release note: `kv.closed_timestamp.lead_for_global_reads_auto_tune.enabled`
can now be used to auto-tune the lead time that ranges global tables use to publish
close timestamps. The `kv.closed_timestamp.lead_for_global_reads_override`
cluster setting takes precedence over this one. If auto-tuning is disabled or
no data is observed, the system falls back to a hardcoded computed lead time.

wenyihu6 committed:
kvserver: cache network RTT for closed timestamp target computation

Previously, during every FlushLockedWithRaftGroup call, we would compute the
maximum network RTT by iterating over all replicas and querying their latencies
from the centralized RPC context. This was expensive as it required accessing
the shared RemoteClocks.Latency callback on this hot path.

This commit introduces a time-based cache for the maximum RTT value. The cache
is refreshed every 5 minutes for stable values, or every 10 seconds if the
previous value was outside acceptable bounds (0-400ms). While this may lead to
slightly less accurate closed timestamp target computations, the performance
benefit outweighs the minor precision loss.

Part of: https://github.com/cockroachdb/cockroach/issues/59680
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142201)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver: use randomized refresh interval to refresh maxReplicaRTT

Previously, all ranges on a node shared the same refresh interval, leading to
simultaneous accesses to the RPC context clock and causing contentions. This
commit introduces a randomized jitter to the refresh interval, ensuring
different ranges on different stores of the same node have varying intervals,
reducing contention and improving performance.

Part of: #59680
Release note: none

wenyihu6 commented:
Closing, going with a different route.
